compiler_opt/rl/generate_bc_trajectories.py

```diff
@@ -529,3 +571,248 @@ def _process_obs(self, curr_obs, sequence_example):
+    horizon = len(seq_example.feature_lists.feature_list['action'].feature)
+    label_list = [label for _ in range(horizon)]
+    add_feature_list(seq_example, label_list, label_name)
+
+  def process_succeeded(
+      self, succeeded, spec_name
+  ) -> Tuple[tf.train.SequenceExample, Dict[str, Union[str, float, int]], Dict[
+      str, Union[str, float, int]]]:
+    seq_example_list = [exploration_res[0] for exploration_res in succeeded]
+    working_dir_list = [(exploration_res[1], exploration_res[2])
+                        for exploration_res in succeeded]
+    seq_example_losses = [exploration_res[3] for exploration_res in succeeded]
+
+    best_policy_idx = np.argmin(seq_example_losses)
+    best_exploration_idx = working_dir_list[best_policy_idx][1]
+
+    # comparator is the last policy in the policy_paths list
+    module_dict_pol = self._profiling_dict(spec_name, seq_example_list[-1][0])
+
+    module_dict_max = self._profiling_dict(
+        spec_name, seq_example_list[best_policy_idx][best_exploration_idx])
+
+    seq_example = seq_example_list[best_policy_idx][best_exploration_idx]
+    best_exploration_idxs = [
+        exploration_res[2] for exploration_res in succeeded
+    ]
+    logging.info('best policy idx: %s, best exploration idxs %s',
+                 best_policy_idx, best_exploration_idxs)
+
+    if self._base_path:
+      # as long as we have one process handles one module this can stay here
+      temp_working_dir_idx = working_dir_list[best_policy_idx][1]
+      temp_working_dir_list = working_dir_list[best_policy_idx][0]
+      temp_working_dir = temp_working_dir_list[temp_working_dir_idx]
+      self._save_binary(self._base_path, spec_name, temp_working_dir)
+
+    return seq_example, module_dict_max, module_dict_pol
+
+  def _profiling_dict(
+      self, module_name: str, feature_list: tf.train.SequenceExample
+  ) -> Dict[str, Union[str, float, int]]:
+    """Return a dictionary for the module containing the name, loss and horizon.
+
+    Args:
+      module_name: name of module
+      feature_list: tf.train.SequenceExample of the compiled module
+
+    Returns:
+      per_module_dict: dictionary containing the name, loss and horizon of
+        compiled module.
+    """
+
+    per_module_dict = {
+        'module_name':
+            module_name,
+        'loss':
+            float(get_loss(feature_list)),
+        'horizon':
+            len(feature_list.feature_lists.feature_list['action'].feature),
+    }
+    return per_module_dict
+
+  def _save_binary(self, base_path, save_path, binary_path):
+    path_head_tail = os.path.split(save_path)
+    path_head = path_head_tail[0]
+    path_tail = path_head_tail[1]
+    save_dir = os.path.join(base_path, path_head)
+    if not os.path.exists(save_dir):
+      os.makedirs(save_dir, exist_ok=True)
+    shutil.copy(
+        os.path.join(binary_path, 'comp_binary'),
+        os.path.join(save_dir, path_tail))
+
+
+@gin.configurable
+class ModuleWorker(worker.Worker):
```

Comment:
It's a bit weird that both `ExplorationWorker` and `ModuleWorker` inherit from `worker.Worker`. It seems like only `ModuleWorker` should inherit from `ExplorationWorker` given this one just wraps `ModuleWorker` and `ExplorationWorker` isn't used on its own?

Reply:
That's probably right. I will make the change.

Reply:
I changed ExplorationWorker to not inherit from Worker. I don't see why ModuleWorker should inherit from ExplorationWorker.

Reply:
Why does ExplorationWorker not need to be a Worker? Put differently, should it have a different suffix then to avoid confusion?

Reply:
I am not sure if it needs to be a worker. I think originally it was used as a worker but then I needed the utility from ModuleWorker and this is why it was inheriting from Worker. I can change the name.

Reply:
so it's not meant to be run in a potentially distributed set up?

Reply:
Okay, that's why I said I am not sure, the way it works right now is ModuleWorker is the one that creates all the distributed tasks and each one is handled by an ExploreModule/ExplorationWorker. Which class needs to inherit from Worker in this case? The way I understood @boomanaiden154's comment was that only ModuleWorker needs to be a Worker, is it the opposite?

Reply:
It was my understanding (originally) that `ExplorationWorker` got distributed and just called into `ModuleWorker`. The class actually getting distributed is what needs to inherit from `worker.Worker`, assuming you're using the distributed training setup.

If we're just creating one `ExplorationWorker` at a time, do we maybe want to rename it to something else like `ExplorationManager` to try and avoid confusion with workers?